manticore/core/plugin.py

```diff
@@ -402,3 +408,256 @@ def will_write_register_callback(self, state, register, value):
+    #: The time that any field of this Descriptor was last updated
+    last_update: datetime = field(default_factory=datetime.now)
+    #: The time at which the on_execution_intermittent callback was last applied to this state. This is when the PC and exec count get updated.
+    last_intermittent_update: typing.Optional[datetime] = None
+    #: The time at which this state was created (or first detected, if the did_enque callback didn't fire for some reason)
+    created_at: datetime = field(default_factory=datetime.now)
+    #: What the state is currently doing (ie waiting for a worker, running, solving, etc.) See enums.StateStatus
+    status: StateStatus = StateStatus.waiting_for_worker
+    #: The last thing a state was doing. Allows us to swap back to this once it finishes solving.
+    _old_status: typing.Optional[StateStatus] = None
+    #: Total number of instruction executions in this state, including those in its parents
+    total_execs: typing.Optional[int] = None
+    #: Number of executions that took place in this state alone, excluding its parents
+    own_execs: typing.Optional[int] = None
+    #: Last program counter (if set)
+    pc: typing.Optional[typing.Any] = None
+    #: Dict mapping field names to the time that field was last updated
+    field_updated_at: typing.Dict[str, datetime] = field(default_factory=dict)
+    #: Message attached to the TerminateState exception that ended this state
+    termination_msg: typing.Optional[str] = None
+
+    def __setattr__(self, key, value):
+        """
+        Force updates the last_updated item any time a field is written to
+        """
+        if key != "last_update":
+            super().__setattr__(key, value)
+        now = datetime.now()
+        # This calls setattr on the _dict_, so it doesn't cause an infinite loop
+        getattr(self, "field_updated_at", {})[key] = now
+        super().__setattr__("last_update", now)
+
+
+class IntrospectionAPIPlugin(Plugin):
+    """
+    Plugin that tracks the movements of States throughout the State lifecycle. Creates a StateDescriptor for each state
+    and stores them in its context, and keeps them up to date whenever a callback registers a change in the State.
+    """
+
+    def create_state(self, state_id: int):
+        """
+        Adds a StateDescriptor to the context in the READY state list
+
+        :param state_id: ID of the state
+        """
+        assert state_id is not None
+        with self.locked_context("manticore_state", dict) as context:
+            context[state_id] = StateDescriptor(state_id=state_id, state_list=StateLists.ready)
+
+    def will_run_callback(self, ready_states: typing.Generator):
+        """
+        Called at the beginning of ManticoreBase.run(). Creates a state descriptor for each of the ready states.
+
+        :param ready_states: Generator that allows us to iterate over the ready states (and modify them if necessary)
+        """
+        for state in ready_states:
```

Comment:
As to elude the generator thing (If that is drained and all that) we could remove it from the event signature and get it from
```self.manticore.ready_states```

Reply:
That would work too, although I'm wary of changing this API since our examples have used `ready_states` as an argument to `will_run_callback` for a long time.

Also, I ended up needing to change the underlying cloning mechanism a bit. Using `itertools.tee` doesn't actually work as intended. While it clones the contents of the generator, the generator is still considered "drained" after the first pass. Since we have code after the `yield`, that code only gets fired once, instead of firing after each plugin has a chance to iterate over the ready states. So, instead, `event.py` now calls `getattr` on the ManticoreBase object so it can get different copies of the `ready_states` generator, instead of just cloning the one.